dashboard/app/models/programming_environment.rb

```diff
@@ -132,10 +132,24 @@ def summarize_for_index
   end
 
   def categories_for_navigation
-    categories.select(&:should_be_in_navigation?).map(&:summarize_for_navigation)
+    Rails.cache.fetch("#{cache_key_with_version}/categories_for_navigation", force: !Script.should_cache?) do
+      categories.select(&:should_be_in_navigation?).map(&:summarize_for_navigation)
+    end
   end
 
   def categories_for_get
     categories.select(&:should_be_in_navigation?).map(&:summarize_for_get)
   end
+
+  def self.get_published_environments_from_cache
+    Rails.cache.fetch("published_programming_environments", force: !Script.should_cache?) do
+      @programming_environments = ProgrammingEnvironment.where(published: true).order(:name).map(&:summarize_for_index)
+    end
+  end
+
+  def self.get_from_cache(name)
+    Rails.cache.fetch("programming_environment/#{name}", force: !Script.should_cache?) do
+      ProgrammingEnvironment.find_by_name(name)
```

Comment:
this looks like a good start!

I think you can improve upon it by doing something like this:
```
@@all_environments ||= get_published_environments_from_cache
@@all_environments.find(...)
```
but it gets tricky because you only want to do this when caching is enabled, or it will make levelbuilder extra slow.

the other opportunity for improvement is to preload this cache.

in both cases I would look to how script caching is done as examples.

Reply:
just clarifying, I think these can be optional, further improvements -- but I would want to use the query count test as a guide for how much work remains to be done here / how much value this would add.

Reply:
I'm curious -- what does the static variable get us here? Just avoiding hitting the cache?

Reply:
one initial query, replacing a query every time a new programming environment is viewed. I am not sure how important this is, unless you are then trying to preload all of them in an initial query. so I think what you have here is probably actually at a local maximum of complexity vs performance  -- we can save further optimization until/unless we decide to preload.

the bigger question in my mind is how much this caching helps vs how much query count is dominated by requests for programming expressions, which it looks like you'll need in order to populate the left nav (if I'm remembering that correctly). 

the way we do it with scripts, we preload all of the script and associated objects. if we copied that here, we might preload programming expressions when we load programming environment. but you can't just do that as an incremental change here, because then you'll slow down levelbuilder. instead you need special logic (like we have for scripts) which says to only pull in the associated models when `should_cache?`. now that I say it, that doesn't sound too hard to add as a conditional though, so don't let me talk you out of it! :-)

Reply:
@bethanyaconnor I am sorry, I clearly missed a bunch of what was in your original PR. it looks like you are already doing the preloading, only you are doing it at the programming expression level, which seems to be working great. given that you are already able to get cached query counts down to zero on the codepaths you care about, I don't think it is important to add the complexity we have with scripts in order to preload them rather than have them get loaded on demand.

Reply:
Great! I have been mulling this comment over the past day or so (I think I have a couple drafts laying around in my tabs  😅 ) but my biggest question was around the fact that these query counts are down to zero. It does sound like we can get some further optimizations by doing some of the things you suggested but I'm not clear that it's actually worth it for the complexity it adds (in my opinion).

I just ran a test where I went to the programming environment page then the programming expression page and the latter request made 4 queries, six fewer than with no caching. So there is improvement that could be made there if we think it's worth it. I don't imagine these pages will cause major performance impacts, so I prefer the simpler solution